src/modules/hip/kernel/non_silent_region_detection.hpp

```diff
@@ -0,0 +1,401 @@
+
+    for (int d = bufferLength >> 1; d > 0; d >>= 1)
+    {
+        __syncthreads();
+        for (int idx = tid; idx < d; idx += hipBlockDim_x)
+        {
+            int ai = offset * (2 * idx + 1) - 1;
+            int bi = offset * (2 * idx + 2) - 1;
+            input[smem_pos(bi)] += input[smem_pos(ai)];
+        }
+        offset <<= 1;
+    }
+
+    if (tid == 0)
+    {
+        int last = bufferLength - 1;
+        input[smem_pos(last)] = 0;
+    }
+
+    for (int d = 1; d < bufferLength; d <<= 1)
+    {
+        offset >>= 1;
+        __syncthreads();
+        for (int idx = tid; idx < d; idx += hipBlockDim_x)
+        {
+            int smem_posA = smem_pos(offset * (2 * idx + 1) - 1);
+            int smem_posB = smem_pos(offset * (2 * idx + 2) - 1);
+            auto t = input[smem_posA];
+            input[smem_posA] = input[smem_posB];
+            input[smem_posB] += t;
+        }
+    }
+    __syncthreads();
+}
+
+// -------------------- Set 1 -  moving mean square compute kernel --------------------
+
+__global__ void moving_mean_square_hip_tensor(float *srcPtr,
+                                              uint nStride,
+                                              float *mmsArr,
+                                              int *srcLengthTensor,
+                                              int outputTileLength,
+                                              int windowLength,
+                                              float windowFactor,
+                                              int inputTileLength)
+{
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+    uint srcLength = srcLengthTensor[id_z];
+    uint batchStride = id_z * nStride;
+    int blockStart = hipBlockIdx_x * outputTileLength;
+
+    if (blockStart >= srcLength)
+        return;
+
+    float *input = srcPtr + batchStride;
+    extern __shared__ float squaredPrefixSum_smem[];
+
+    float *inBlockPtr = srcPtr + batchStride + blockStart;
+    float *outBlockPtr = mmsArr + batchStride + blockStart;
+    int validOutputTileLength = std::min<int>(outputTileLength, srcLength - blockStart);
+    float *extendedBlockStart = inBlockPtr - windowLength;
+    float *extendedBlockEnd = inBlockPtr + validOutputTileLength;
+
+    // load input data to shared memory
+    for(int pos = hipThreadIdx_x; pos < inputTileLength; pos += hipBlockDim_x)
+    {
+        float val = 0.0f;
+        auto extendedBlockPtr = extendedBlockStart + pos;
+        if (extendedBlockPtr >= input && extendedBlockPtr < extendedBlockEnd)
+            val = *extendedBlockPtr;
+        squaredPrefixSum_smem[smem_pos(pos)] = square(val);
+    }
+
+    // compute prefix sum
+    compute_prefix_sum(squaredPrefixSum_smem, inputTileLength);
+
+    // compute the mms value here
+    for(int pos = hipThreadIdx_x; pos < validOutputTileLength; pos += hipBlockDim_x)
+    {
+        float x = inBlockPtr[pos];
```

Comment:
We need to run pos for each nth thread jumping across blocks? Any way this can be avoided due to the inBlockPtr[pos] global memaccess?

Reply:
since we need a prefix sum for window length values and it has be reset for every reset interval
we cannot avoid this here